scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -304,24 +304,34 @@
-                                                   (log/warn "In compute cluster" name ", pod" pod-name
-                                                             "in a weird cook expected state:"
-                                                             (prepare-cook-expected-state-dict-for-logging cook-expected-state-dict)
-                                                             "and k8s actual state"
-                                                             (prepare-k8s-actual-state-dict-for-logging k8s-actual-state-dict))
-                                                   ; TODO: Avoid a race. Say a launch occurs, followed by a kill, followed by a watch update.
-                                                   ; Before the watch update, we'll think the k8s-actual-state is missing.
-                                                   ; We see (:killed,:missing), with nothing to do, so we move to (;missing,:missing)
-                                                   ; Then the watch update occurs and we see (:missing,:starting), log a weird state,
-                                                   ; and kill it off. We want to avoid that path by doing an opportunistic kill here.
-                                                   ; That way we if we're seeing a stale :missing, we'll kill it anyways.
-                                                   ; We can't do that now because pod is nil, and kill-pod requires non-nil pod.
-
-                                                   (handle-pod-killed compute-cluster pod-name))
+                                        :missing
+                                        ; There's a race where we can launch then kill, when the kill arrives after
+                                        ; the launch, but before the watch updates k8s-actual-state.
+                                        ; (k8s-actual-state isn't the actual state, because of this watch lag)
+                                        ; When that happens, we will have an actual state of :missing.
+                                        ; If we did nothing, we'd log this as a weird state, then when the watch
+                                        ; shows up, we'd see (:missing, :starting) and log that as a weird state too.
+                                        ;
+                                        ; So, a better approach. If we detect a :missing,:killed, then we opportunistically
+                                        ; try to kill the pod. This is why update-cook-expected-state saves :launch-pod,
+                                        ; so its available here.
+                                        (if-let [pod (some-> cook-expected-state-dict :launch-pod :pod)]
+                                          (do (log/info "In compute cluster" name ", opportunistically killing" pod-name "because of potential race where kill arrives before the watch responds to the launch")
```

Comment:
The `log/info` should start on the next line.

Also, the line seems too long.

Reply:
Ok. Fixed.